test/max_message_test.c

```diff
@@ -0,0 +1,229 @@
+	(void)friend_number;
+	(void)message;
+	(void)user_data;
+	last_received_length = length;
+}
+
+static void test_receipt_cb(uint32_t friend_number, const uint64_t receipt_id,
+			    void *user_data)
+{
+	(void)friend_number;
+	(void)user_data;
+	(void)receipt_id;
+}
+
+static void test_neg_cb(uint32_t friend_number, bool compatible,
+			void *user_data)
+{
+	(void)friend_number;
+	(void)compatible;
+	(void)user_data;
+}
+
+uint64_t test_tox_extension_messages_append_no_size_check(
+	struct ToxExtensionMessages *extension,
+	struct ToxExtPacketList *packet_list, uint8_t const *data, size_t size,
+	uint32_t friend_id, enum Tox_Extension_Messages_Error *err)
+{
+	(void)friend_id;
+
+	uint8_t const *end = data + size;
+	uint8_t const *next_chunk = data;
+	bool first_chunk = true;
+	uint64_t receipt_id = extension->next_receipt_id++;
+	do {
+		uint8_t extension_data[TOXEXT_MAX_SEGMENT_SIZE];
+		size_t size_for_chunk;
+		next_chunk = tox_extension_messages_chunk(
+			first_chunk, next_chunk, end - next_chunk, receipt_id,
+			extension_data, &size_for_chunk);
+		first_chunk = false;
+
+		toxext_segment_append(packet_list, extension->extension_handle,
+				      extension_data, size_for_chunk);
+	} while (end > next_chunk);
+
+	if (err) {
+		*err = TOX_EXTENSION_MESSAGES_SUCCESS;
+	}
+	return receipt_id;
+}
+
+void test_unnegotiated_size(struct ToxExtUser *user_a,
+			    struct ToxExtUser *user_b,
+			    struct ToxExtensionMessages *ext_a,
+			    struct ToxExtensionMessages *ext_b)
+{
+	enum Tox_Extension_Messages_Error err;
+	tox_extension_messages_get_max_sending_size(ext_a, user_b->tox_user.id,
+						    &err);
+	assert(err == TOX_EXTENSION_MESSAGES_INVALID_ARG);
+	tox_extension_messages_get_max_sending_size(ext_b, user_a->tox_user.id,
+						    &err);
+	assert(err == TOX_EXTENSION_MESSAGES_INVALID_ARG);
+}
+
+void test_negotiated_size(struct ToxExtUser *user_a, struct ToxExtUser *user_b,
+			  struct ToxExtensionMessages *ext_a,
+			  struct ToxExtensionMessages *ext_b)
+{
+	enum Tox_Extension_Messages_Error err;
+	uint64_t max_size = tox_extension_messages_get_max_sending_size(
+		ext_a, user_b->tox_user.id, &err);
+	assert(err == TOX_EXTENSION_MESSAGES_SUCCESS);
+	assert(max_size ==
+	       TOX_EXTENSION_MESSAGES_DEFAULT_MAX_RECEIVING_MESSAGE_SIZE);
+	max_size = tox_extension_messages_get_max_sending_size(
+		ext_b, user_a->tox_user.id, &err);
+	assert(err == TOX_EXTENSION_MESSAGES_SUCCESS);
+	assert(max_size == a_max_size);
+}
+
+void test_sending_over_max(struct ToxExtUser *user_a, struct ToxExtUser *user_b,
+			   struct ToxExtensionMessages *ext_b)
+{
+	enum Tox_Extension_Messages_Error err;
+	struct ToxExtPacketList *packet_list =
+		toxext_packet_list_create(user_b->toxext, user_a->tox_user.id);
+	uint8_t buffer[a_max_size + 1];
+	tox_extension_messages_append(ext_b, packet_list, buffer,
+				      a_max_size + 1, user_a->tox_user.id,
+				      &err);
+	assert(err == TOX_EXTENSION_MESSAGES_INVALID_ARG);
+	toxext_send(packet_list);
```

Comment:
This is pretty weird because this should be an empty packet list. Indicates that we should probably add a toxext_packetlist_free that allows the caller to abort the send. That being said I don't think there's anything wrong here so I'm not going to bother requesting any changes here.

We _could_ check that the last_received_message_size doesn't change but I'm not going to block this on that since this has been in review for so long and I don't think it's wrong.